### PR TITLE
New settings UI

### PR DESCRIPTION
**Pull request by jaymckinney-enavate**
The first half of the new settings enhancement.
Will commit module setting implementations when they are finished.

https://dev.azure.com/EPS-Product/EPS-Product/_sprints/taskboard/EPS-Product%20Team/EPS-Product/2022/Sprint%2015?workitem=749

### CODE REVIEW

**Review by ToddE123 — COMMENTED** [review]
I think checks should be moved upstream and out of the execution procedures. Thoughts?

File: Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al [hunk 1]
```diff
@@ -106,10 +106,10 @@ codeunit 4025 "GP Cloud Migration"
         end;
 
         Commit();
-        if GPCompanyMigrationSettings.Get(CompanyName()) then begin
```
[the remark this answers is not in the document]
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al:109]: The dimension fields were moved to the new 'additional settings' table.

File: Apps/W1/HybridGP/app/src/tables/GPCompanyMigrationSettings.table.al [hunk 2]
```diff
@@ -18,62 +18,6 @@ table 4044 "GP Company Migration Settings"
             CalcFormula = lookup("Hybrid Company".Replicate where(Name = field(Name)));
         }
 
-        field(4; "Global Dimension 1"; Text[30])
```
[the remark this answers is not in the document]
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/tables/GPCompanyMigrationSettings.table.al:21]: I see, that makes sense. I could add them back and have them updated from the new UI.

File: Apps/W1/HybridGP/app/src/Migration/Customers/GPCustomerMigrator.codeunit.al [hunk 3]
```diff
@@ -12,10 +12,15 @@ codeunit 4018 "GP Customer Migrator"
     [EventSubscriber(ObjectType::Codeunit, Codeunit::"Customer Data Migration Facade", 'OnMigrateCustomer', '', true, true)]
     procedure OnMigrateCustomer(var Sender: Codeunit "Customer Data Migration Facade"; RecordIdToMigrate: RecordId)
     var
+        GPCompanyAdditionalSettings: Record "GP Company Additional Settings";
         GPCustomer: Record "GP Customer";
     begin
         if RecordIdToMigrate.TableNo() <> Database::"GP Customer" then
             exit;
+
+        if not GPCompanyAdditionalSettings.GetReceivablesModuleEnabled() then
```
[the remark this answers is not in the document]
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/Migration/Customers/GPCustomerMigrator.codeunit.al:21]: Good to know, I'll move them upstream. Thanks.

File: Apps/W1/HybridGP/app/src/Migration/Customers/GPCustomerMigrator.codeunit.al [hunk 4]
```diff
@@ -57,6 +67,9 @@ codeunit 4018 "GP Customer Migrator"
         if RecordIdToMigrate.TableNo() <> Database::"GP Customer" then
             exit;
 
+        if not GPCompanyAdditionalSettings.GetReceivablesModuleEnabled() then
```
[the remark this answers is not in the document]
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/Migration/Customers/GPCustomerMigrator.codeunit.al:70]: Will move this one too.

File: Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al:109]: I saw your later comment, I'll move them back and have the new settings UI update the old fields.

File: Apps/W1/HybridGP/app/src/Migration/Items/GPItemMigrator.codeunit.al [hunk 5]
```diff
@@ -60,7 +60,7 @@ codeunit 4019 "GP Item Migrator"
             MigrateItemInventoryPostingGroup(GPItem, Sender);
     end;
 
-    procedure MigrateItemInventoryPostingGroup(GPItem: Record "GP Item"; var Sender: Codeunit "Item Data Migration Facade")
```
Comment (ToddE123) [inline Apps/W1/HybridGP/app/src/Migration/Items/GPItemMigrator.codeunit.al:63]: why the change to the procedure's definition? Is this allowed?

File: Apps/W1/HybridGP/app/src/Migration/Support/CheckBooks/GPCheckbookMigrator.codeunit.al [hunk 6]
```diff
@@ -11,6 +11,9 @@ codeunit 40025 "GP Checkbook Migrator"
         GPCompanyAdditionalSettings: Record "GP Company Additional Settings";
         MigrateInactiveCheckbooks: Boolean;
     begin
+        if not GPCompanyAdditionalSettings.GetBankModuleEnabled() then
```
Comment (ToddE123) [inline Apps/W1/HybridGP/app/src/Migration/Support/CheckBooks/GPCheckbookMigrator.codeunit.al:14]: I don't think this is needed as you have the same check in the helper functions which calls this procedure.

File: Apps/W1/HybridGP/app/src/Migration/Support/GPConfiguration.table.al [hunk 7]
```diff
@@ -104,14 +104,16 @@ table 4024 "GP Configuration"
...
+        GPCompanyAdditionalSettings: Record "GP Company Additional Settings";
     begin
         exit(
-                "CheckBooks Created" and
-                "Open Purchase Orders Created" and
                 "Fiscal Periods Created" and
-                "Vendor EFT Bank Acc. Created" and
-                "Vendor Classes Created" and
-                "Customer Classes Created"
+                (not GPCompanyAdditionalSettings.GetBankModuleEnabled() or "CheckBooks Created") and
```
Comment (ToddE123) [inline Apps/W1/HybridGP/app/src/Migration/Support/GPConfiguration.table.al:112]: glad you wrote this, my BC and\or statement writing could use some work...I would have just used a number of exit(value) statements. ;)

File: Apps/W1/HybridGP/app/src/Migration/Support/PurchaseOrders/GPPOPHeader.table.al [hunk 8]
```diff
@@ -797,6 +798,9 @@ table 40102 "GP POPPOHeader"
         PurchaseDocumentStatus: Enum "Purchase Document Status";
         CountryCode: Code[10];
     begin
+        if not GPCompanyAdditionalSettings.GetMigrateOpenPOs() then
```
Comment (ToddE123) [inline Apps/W1/HybridGP/app/src/Migration/Support/PurchaseOrders/GPPOPHeader.table.al:801]: is this needed? You have the same check in the helper functions which calls this procedure.

File: Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al [hunk 9]
```diff
@@ -111,16 +107,13 @@ codeunit 4025 "GP Cloud Migration"
...
-        CreateDataMigrationStatusRecords(Database::"Vendor", GPVendor.Count(), 4096, 4022);
-        CreateDataMigrationStatusRecords(Database::"Item", GPItem.Count(), 4095, 4019);
+        CreateAllDataMigrationStatusRecords(DataMigrationEntity);
 
         Session.LogMessage('0000BBI', StartMigrationMsg, Verbosity::Normal, DataClassification::SystemMetadata, TelemetryScope::ExtensionPublisher, 'Category', HelperFunctions.GetTelemetryCategory());
         DataMigrationFacade.StartMigration(HelperFunctions.GetMigrationTypeTxt(), FALSE);
     end;
 
-    local procedure CreateDataMigrationStatusRecords(DestinationTableID: Integer; NumberOfRecords: Integer; StagingTableID: Integer; CodeunitToRun: Integer)
+    procedure CreateDataMigrationStatusRecords(DestinationTableID: Integer; NumberOfRecords: Integer; StagingTableID: Integer; CodeunitToRun: Integer)
```
Comment (ToddE123) [inline Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al:116]: why the change from local procedure?

File: Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al [hunk 10]
```diff
@@ -135,17 +128,45 @@ codeunit 4025 "GP Cloud Migration"
         DataMigrationStatus.Insert()
     end;
 
-    local procedure CreateDataMigrationEntites(var DataMigrationEntity: Record "Data Migration Entity"): Boolean
+    procedure CreateDataMigrationEntites(var DataMigrationEntity: Record "Data Migration Entity"): Boolean
```
Comment (ToddE123) [inline Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al:131]: why the change from local procedure?

File: Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al [hunk 11]
```diff
@@ -111,16 +107,13 @@ codeunit 4025 "GP Cloud Migration"
             HelperFunctions.UpdateGlobalDimensionNo();
         end;
 
-        CreateDataMigrationStatusRecords(Database::"G/L Account", GPAccount.Count(), 4090, 4017);
-        CreateDataMigrationStatusRecords(Database::"Customer", GPCustomer.Count(), 4093, 4018);
-        CreateDataMigrationStatusRecords(Database::"Vendor", GPVendor.Count(), 4096, 4022);
-        CreateDataMigrationStatusRecords(Database::"Item", GPItem.Count(), 4095, 4019);
+        CreateAllDataMigrationStatusRecords(DataMigrationEntity);
```
Comment (ToddE123) [inline Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al:110]: the 'All' in the name infers we are creating all records when we are not. Suggest removing 'All' from the procedure name.

File: Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al [hunk 12]
```diff
@@ -135,17 +128,45 @@ codeunit 4025 "GP Cloud Migration"
...
+        if GPCompanyAdditionalSettings.GetPayablesModuleEnabled() then
+            DataMigrationEntity.InsertRecord(Database::Vendor, HelperFunctions.GetNumberOfVendors());
+
+        if GPCompanyAdditionalSettings.GetInventoryModuleEnabled() then
+            DataMigrationEntity.InsertRecord(Database::Item, HelperFunctions.GetNumberOfItems());
+
         exit(true);
     end;
 
+    procedure CreateAllDataMigrationStatusRecords(var DataMigrationEntity: Record "Data Migration Entity")
```
Comment (ToddE123) [inline Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al:150]: can this be a local procedure too?

File: Apps/W1/HybridGP/app/src/pages/GPCompanyAdditionalSettingsList.Page.al [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    layout
+    {
+        area(content)
+        {
+            repeater(General)
+            {
+                field(Name; Rec.Name)
+                {
+                    Caption = 'Company';
+                    ToolTip = 'Specifies the value of the Name field.';
```
Comment (ToddE123) [inline Apps/W1/HybridGP/app/src/pages/GPCompanyAdditionalSettingsList.Page.al:20]: Should 'Company' be used somewhere in this tooltip?

**Review by ToddE123 — COMMENTED** [review]
Not sure exactly I was looking for, but I didn't notice anything here to stop a user from running the migration before using this new UI. Will that be in the core extension or another place or not at all?

File: Apps/W1/HybridGP/app/src/Migration/Support/CheckBooks/GPCheckbookMigrator.codeunit.al (hunk 6 above)
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/Migration/Support/CheckBooks/GPCheckbookMigrator.codeunit.al:14]: Yes, I missed this one.

File: Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al (hunk 11 above)
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/codeunits/GPCloudMigration.codeunit.al:110]: Good point. I'm renaming to CreateConfiguredDataMigrationStatusRecords since that better describes what it's for.

File: Apps/W1/HybridGP/app/src/pages/GPCompanyAdditionalSettingsList.Page.al (hunk 13 above)
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/pages/GPCompanyAdditionalSettingsList.Page.al:20]: Ha! Yes, I need to update this.

File: Apps/W1/HybridGP/app/src/Migration/Support/HelperFunctions.codeunit.al [hunk 14]
```diff
@@ -69,6 +69,12 @@ Codeunit 4037 "Helper Functions"
         CloudMigrationTok: Label 'CloudMigration', Locked = true;
         GeneralTemplateNameTxt: Label 'GENERAL', Locked = true;
 
+    [EventSubscriber(ObjectType::Codeunit, CodeUnit::"Hybrid Cloud Management", 'OnBeforeShowProductSpecificSettingsPageStep', '', true, true)]
```
[the remark this answers is not in the document]
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/Migration/Support/HelperFunctions.codeunit.al:72]: Makes sense. I think I found the one you meant, let me know if I'm incorrect. Moved to HybridGPManagement.codeunit.al

File: Apps/W1/HybridGP/app/src/pages/GPCompanyAdditionalSettingsList.Page.al [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+            }
+        }
+    }
+
+    local procedure PromptForAutoPopulateDimension(DimensionNumber: Integer; DimensionLabel: Text[30])
+    var
+        GPCompanyAdditionalSettings: Record "GP Company Additional Settings";
+    begin
+        // Only prompt if there is more than 1 company being migrated. (Global settings record included in the count)
+        if GPCompanyAdditionalSettings.Count() < 3 then
```
[the remark this answers is not in the document]
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/pages/GPCompanyAdditionalSettingsList.Page.al:130]: This logic is being replaced by the 'Set All Dimensions' dialog (next commit).

File: Apps/W1/HybridGP/app/src/pages/HybridCloudWizardExtension.PageExt.al [hunk 16]
```diff
@@ -1,39 +1,30 @@
...
-                        UpdatePropagation = Both;
-                    }
-                }
-                group("GPMigrationSettings.2.0")
-                {
-                    ShowCaption = false;
-                    group("GPMigrationSettings.2.1")
-                    {
-                        ShowCaption = false;
-                        InstructionalText = 'Select the two segments from Dynamics GP you would like as the global dimensions. The remaining segments will automatically be set up as shortcut dimensions.';
```
[the remark this answers is not in the document]
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/pages/HybridCloudWizardExtension.PageExt.al:28]: I don't recall ever seeing this. Was it shown in the wizard?

[the remark this answers is not in the document]
Comment (ToddE123) [reply Apps/W1/HybridGP/app/src/pages/HybridCloudWizardExtension.PageExt.al:28]: It was at the bottom of the "Cloud Migration Setup" wizard when you selected your dimensions.

[the remark this answers is not in the document]
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/app/src/pages/HybridCloudWizardExtension.PageExt.al:28]: Found it and I've added it to both the main settings page and the 'Set All Dimensions' dialog.

**Comment by jaymckinney-enavate** [comment]
Replaced with a newer PR which incorporates the new Oldest GL feature.
